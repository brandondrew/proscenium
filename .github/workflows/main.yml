name: Ruby

on:
  push:
    branches:
      - master

  pull_request:

jobs:
  release:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v3

      - name: Set up Ruby
        uses: ruby/setup-ruby@v1
        with:
          ruby-version: 3.1

      - uses: denoland/setup-deno@v1.0.0
        with:
          deno-version: v1.x

      - name: Compile esbuild
        run: deno task compile:esbuild:macm

      - name: Install gem dependencies
        run: bundle install

      - name: Build native gem
        env:
          PROFILE: release
        run: bundle exec rake build

      - uses: actions/upload-artifact@v3
        with:
          name: artifacts
          path: bin/esbuild-*
