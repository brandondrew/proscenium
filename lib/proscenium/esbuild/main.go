package main

// go build -buildmode=c-shared -o main.so main.go

import "C"
import "os"
import "strings"
import "encoding/json"

// import "fmt"
import "github.com/evanw/esbuild/pkg/api"

//export transform
func transform(input *C.char) *C.char {
	inputString := C.GoString(input)
	result := api.Transform(inputString, api.TransformOptions{
		Loader: api.LoaderJS,
	})

	s := string(result.Code)

	return C.CString(s)
}

var envPlugin = api.Plugin{
	Name: "env",
	Setup: func(build api.PluginBuild) {
		// Intercept import paths called "env" so esbuild doesn't attempt
		// to map them to a file system location. Tag them with the "env-ns"
		// namespace to reserve them for this plugin.
		build.OnResolve(api.OnResolveOptions{Filter: `^env$`},
			func(args api.OnResolveArgs) (api.OnResolveResult, error) {
				return api.OnResolveResult{
					Path:      args.Path,
					Namespace: "env-ns",
				}, nil
			})

		// Load paths tagged with the "env-ns" namespace and behave as if
		// they point to a JSON file containing the environment variables.
		build.OnLoad(api.OnLoadOptions{Filter: `.*`, Namespace: "env-ns"},
			func(args api.OnLoadArgs) (api.OnLoadResult, error) {
				mappings := make(map[string]string)
				for _, item := range os.Environ() {
					if equals := strings.IndexByte(item, '='); equals != -1 {
						mappings[item[:equals]] = item[equals+1:]
					}
				}
				bytes, err := json.Marshal(mappings)
				if err != nil {
					return api.OnLoadResult{}, err
				}
				contents := string(bytes)
				return api.OnLoadResult{
					Contents: &contents,
					Loader:   api.LoaderJSON,
				}, nil
			})
	},
}

//export build
func build(entrypoint *C.char) *C.char {
	entrypointString := C.GoString(entrypoint)

	result := api.Build(api.BuildOptions{
		EntryPoints: []string{entrypointString},
		Bundle:      true,
		Plugins:     []api.Plugin{envPlugin},
		Write:       false,
	})

	outputString := string(result.OutputFiles[0].Contents)
	return C.CString(outputString)
}

func main() {
	// fmt.Println(transform())
	// fmt.Printf("%s", build("input.ts"))
}
