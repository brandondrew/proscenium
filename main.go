package main

/*
struct Result {
	int success;
	char* response;
};
*/
import "C"

import (
	"encoding/json"
	"joelmoss/proscenium/internal/builder"
	"joelmoss/proscenium/internal/resolver"
	"joelmoss/proscenium/internal/types"
	"path"
	"strings"
)

// Build the given `path` in the `root`.
//
//   - path - The path to build relative to `root`.
//   - root - The working directory.
//   - env - The environment (1 = development, 2 = test, 3 = production)
//   - importMap - Path to the import map relative to `root`.
//   - debug
//
//export build
func build(filepath *C.char, root *C.char, env C.uint, importMap *C.char, debug bool) C.struct_Result {
	types.Env = types.Environment(env)

	pathStr := C.GoString(filepath)

	result := builder.Build(builder.BuildOptions{
		Path:          pathStr,
		Root:          C.GoString(root),
		ImportMapPath: C.GoString(importMap),
		Debug:         debug,
	})

	if len(result.Errors) != 0 {
		j, err := json.Marshal(result.Errors[0])
		if err != nil {
			return C.struct_Result{C.int(0), C.CString(string(err.Error()))}
		}

		return C.struct_Result{C.int(0), C.CString(string(j))}
	}

	contents := string(result.OutputFiles[0].Contents)

	isSourceMap := strings.HasSuffix(pathStr, ".map")
	if isSourceMap {
		return C.struct_Result{C.int(1), C.CString(contents)}
	}

	sourcemapUrl := path.Base(pathStr)
	if strings.HasSuffix(result.OutputFiles[0].Path, ".css") {
		contents += "/*# sourceMappingURL=" + sourcemapUrl + ".map */"
	} else {
		contents += "//# sourceMappingURL=" + sourcemapUrl + ".map"
	}

	return C.struct_Result{C.int(1), C.CString(contents)}
}

// Resolve the given `path` relative to the `root`.
//
//   - path - The path to build relative to `root`.
//   - root - The working directory.
//   - env - The environment (1 = development, 2 = test, 3 = production)
//   - importMap - Path to the import map relative to `root`.
//
//export resolve
func resolve(path *C.char, root *C.char, env C.uint, importMap *C.char) C.struct_Result {
	types.Env = types.Environment(env)

	resolvedPath, err := resolver.Resolve(resolver.Options{
		Path:          C.GoString(path),
		Root:          C.GoString(root),
		ImportMapPath: C.GoString(importMap),
	})
	if err != nil {
		return C.struct_Result{C.int(0), C.CString(string(err.Error()))}
	}

	return C.struct_Result{C.int(1), C.CString(resolvedPath)}
}

func main() {}
