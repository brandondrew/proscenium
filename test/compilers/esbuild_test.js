import { assertRejects } from 'testing/asserts.ts'
import { assertSnapshot } from 'testing/snapshot.ts'
import { beforeEach, describe, it } from 'testing/bdd.ts'
import { join } from 'std/path/mod.ts'

import main from '../../lib/proscenium/compilers/esbuild.js'
import ArgumentError from '../../lib/proscenium/compilers/esbuild/argument_error.js'

const root = join(Deno.cwd(), 'test', 'internal')
const lightningcssBin = join(Deno.cwd(), 'bin', 'lightningcss')

describe('compilers/esbuild', () => {
  beforeEach(() => {
    Deno.env.set('RAILS_ENV', 'test')
    Deno.env.set('PROSCENIUM_TEST', 'test')
  })

  it('throws without any arguments', async () => {
    await assertRejects(async () => await main(), ArgumentError, 'pathRequired')
  })

  it('throws without root option', async () => {
    await assertRejects(async () => await main('**/*.js'), ArgumentError, 'rootRequired')
  })

  it('throws without lightningcssBin option', async () => {
    await assertRejects(
      async () => await main('**/*.js', { root: 'foo/bar' }),
      ArgumentError,
      'lightningcssBinRequired:'
    )
  })

  it('Successful JSX build', async t => {
    const result = await main('lib/component.jsx', { root, lightningcssBin })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })

  it('Import bare module', async t => {
    const result = await main('lib/import_node_module.js', { root, lightningcssBin })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })

  it('invalid URL', async t => {
    const result = await main('url:https://unpkg.com/open-props@1.5.3/red.min.css', {
      root,
      lightningcssBin
    })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })

  it('allows returns error on unknown bare module', async t => {
    const result = await main('lib/import_unknown_node_module.js', { root, lightningcssBin })

    await assertSnapshot(t, result)
  })

  it('resolves nested node modules', async t => {
    const result = await main('node_modules/@react-aria/button', { root, lightningcssBin })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })

  it('Import relative module', async t => {
    const result = await main('lib/import_relative_module.js', { root, lightningcssBin })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })

  it('Import absolute module', async t => {
    const result = await main('lib/import_absolute_module.js', { root, lightningcssBin })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })

  it('Import remote module', async t => {
    const result = await main('lib/import_remote_module.js', { root, lightningcssBin })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })

  it('npm module with relative import', async t => {
    const result = await main('lib/npm_module_with_relative_import.js', { root, lightningcssBin })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })

  it('Import relative module without extension', async t => {
    const result = await main('lib/import_relative_module_without_extension.js', {
      root,
      lightningcssBin
    })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })

  it('Import absolute module without extension', async t => {
    const result = await main('lib/import_absolute_module_without_extension.js', {
      root,
      lightningcssBin
    })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })

  it('Import css module from JS', async t => {
    const result = await main('lib/import_css_module.js', {
      root,
      lightningcssBin,
      cssMixinPaths: [join(root, 'lib')]
    })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })

  it('de-dupes side loaded ViewComponent default stylesheet - regular', async t => {
    const result = await main('app/components/basic_react_component.jsx', {
      root,
      lightningcssBin
    })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })

  it('de-dupes side loaded ViewComponent default stylesheet - sidecar', async t => {
    const result = await main('app/components/basic_react/component.jsx', {
      root,
      lightningcssBin
    })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })

  it('Import css from JS', async t => {
    const result = await main('lib/import_css.js', {
      root,
      lightningcssBin,
      cssMixinPaths: [join(root, 'lib')]
    })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })

  it('Import css from jsx', async t => {
    const result = await main('lib/import_css.jsx', {
      root,
      lightningcssBin,
      cssMixinPaths: [join(root, 'lib')]
    })

    await assertSnapshot(t, new TextDecoder().decode(result))
  })

  describe('postcss', () => {
    it('supports mixins', async t => {
      const result = await main('lib/with_mixins.css', {
        root,
        lightningcssBin,
        cssMixinPaths: [join(root, 'lib')]
      })

      await assertSnapshot(t, new TextDecoder().decode(result))
    })

    it('cssMixinPaths option', async t => {
      const result = await main('lib/with_mixins_from_alternative_path.css', {
        root,
        lightningcssBin,
        cssMixinPaths: [join(root, 'lib'), join(root, 'config')]
      })

      await assertSnapshot(t, new TextDecoder().decode(result))
    })
  })
})
