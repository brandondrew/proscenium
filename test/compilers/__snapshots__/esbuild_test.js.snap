export const snapshot = {};

snapshot[`compilers/esbuild > Successful JSX build 1`] = `
'// lib/component.jsx
import { jsx } from "/node_modules/react/jsx-runtime.js";
var Component = () => {
  return /* @__PURE__ */ jsx("div", { children: "Hello" });
};
var component_default = Component;
export {
  component_default as default
};
//# sourceMappingURL=component.jsx.map
'
`;

snapshot[`compilers/esbuild > Import bare module 1`] = `
'// lib/import_node_module.js
import myPackage from "/node_modules/mypackage/index.js";
import isIp from "/node_modules/is-ip/index.js";
var import_node_module_default = () => {
  myPackage();
  isIp("?");
};
export {
  import_node_module_default as default
};
//# sourceMappingURL=import_node_module.js.map
'
`;

snapshot[`compilers/esbuild > allows returns error on unknown bare module 1`] = `
{
  detail: undefined,
  id: "",
  location: {
    column: 20,
    file: "lib/import_unknown_node_module.js",
    length: 9,
    line: 1,
    lineText: "import unknown from 'unknown'",
    namespace: "",
    suggestion: "",
  },
  notes: [
    {
      location: null,
      text: 'You can mark the path "unknown" as external to exclude it from the bundle, which will remove this error.',
    },
  ],
  pluginName: "resolve",
  text: 'Could not resolve "unknown"',
}
`;

snapshot[`compilers/esbuild > resolves nested node modules 1`] = `
'// node_modules/@react-aria/button/dist/module.js
import { mergeProps as \$cE0pI\$mergeProps, filterDOMProps as \$cE0pI\$filterDOMProps, chain as \$cE0pI\$chain } from "/node_modules/.pnpm/@react-aria+button@3.6.3_react@18.2.0/node_modules/@react-aria/utils/dist/module.js";
import { useFocusable as \$cE0pI\$useFocusable } from "/node_modules/.pnpm/@react-aria+button@3.6.3_react@18.2.0/node_modules/@react-aria/focus/dist/module.js";
import { usePress as \$cE0pI\$usePress } from "/node_modules/.pnpm/@react-aria+button@3.6.3_react@18.2.0/node_modules/@react-aria/interactions/dist/module.js";
function \$701a24aa0da5b062\$export\$ea18c227d4417cc3(props, ref) {
  let {
    elementType = "button",
    isDisabled,
    onPress,
    onPressStart,
    onPressEnd,
    onPressChange,
    preventFocusOnPress,
    allowFocusWhenDisabled,
    onClick: deprecatedOnClick,
    href,
    target,
    rel,
    type = "button"
  } = props;
  let additionalProps;
  if (elementType === "button")
    additionalProps = {
      type,
      disabled: isDisabled
    };
  else
    additionalProps = {
      role: "button",
      tabIndex: isDisabled ? void 0 : 0,
      href: elementType === "a" && isDisabled ? void 0 : href,
      target: elementType === "a" ? target : void 0,
      type: elementType === "input" ? type : void 0,
      disabled: elementType === "input" ? isDisabled : void 0,
      "aria-disabled": !isDisabled || elementType === "input" ? void 0 : isDisabled,
      rel: elementType === "a" ? rel : void 0
    };
  let { pressProps, isPressed } = \$cE0pI\$usePress({
    onPressStart,
    onPressEnd,
    onPressChange,
    onPress,
    isDisabled,
    preventFocusOnPress,
    ref
  });
  let { focusableProps } = \$cE0pI\$useFocusable(props, ref);
  if (allowFocusWhenDisabled)
    focusableProps.tabIndex = isDisabled ? -1 : focusableProps.tabIndex;
  let buttonProps = \$cE0pI\$mergeProps(focusableProps, pressProps, \$cE0pI\$filterDOMProps(props, {
    labelable: true
  }));
  return {
    isPressed,
    buttonProps: \$cE0pI\$mergeProps(additionalProps, buttonProps, {
      "aria-haspopup": props["aria-haspopup"],
      "aria-expanded": props["aria-expanded"],
      "aria-controls": props["aria-controls"],
      "aria-pressed": props["aria-pressed"],
      onClick: (e) => {
        if (deprecatedOnClick) {
          deprecatedOnClick(e);
          console.warn("onClick is deprecated, please use onPress");
        }
      }
    })
  };
}
function \$55f54f7887471b58\$export\$51e84d46ca0bc451(props, state, ref) {
  const { isSelected } = state;
  const { isPressed, buttonProps } = \$701a24aa0da5b062\$export\$ea18c227d4417cc3({
    ...props,
    onPress: \$cE0pI\$chain(state.toggle, props.onPress)
  }, ref);
  return {
    isPressed,
    buttonProps: \$cE0pI\$mergeProps(buttonProps, {
      "aria-pressed": isSelected
    })
  };
}
export {
  \$701a24aa0da5b062\$export\$ea18c227d4417cc3 as useButton,
  \$55f54f7887471b58\$export\$51e84d46ca0bc451 as useToggleButton
};
//# sourceMappingURL=button.js.map
'
`;

snapshot[`compilers/esbuild > Import relative module 1`] = `
'// lib/import_relative_module.js
import "/lib/foo.js";
//# sourceMappingURL=import_relative_module.js.map
'
`;

snapshot[`compilers/esbuild > Import absolute module 1`] = `
'// lib/import_absolute_module.js
import "/lib/foo.js";
//# sourceMappingURL=import_absolute_module.js.map
'
`;

snapshot[`compilers/esbuild > Import remote module 1`] = `
'// lib/import_remote_module.js
import emailRegex from "/url:https%3A%2F%2Fesm.sh%2Femail-regex";
//# sourceMappingURL=import_remote_module.js.map
'
`;

snapshot[`compilers/esbuild > import map > from json 1`] = `
'// lib/import_map.js
import axios from "/url:https%3A%2F%2Fcdnjs.cloudflare.com%2Fajax%2Flibs%2Faxios%2F0.24.0%2Faxios.min.js";
import { createElement } from "/url:https%3A%2F%2Fesm.sh%2Freact%4018.2.0";
import foo from "/lib/foo.js";
axios();
foo();
createElement();
import("/url:https%3A%2F%2Fesm.sh%2Fsolid%400.2.1");
//# sourceMappingURL=import_map.js.map
'
`;

snapshot[`compilers/esbuild > import map > resolves imports from a node_module 1`] = `
'// node_modules/is-ip/index.js
import ipRegex from "/url:https%3A%2F%2Fesm.sh%2Fip-regex";
import { isMatch } from "/node_modules/.pnpm/is-ip@5.0.0/node_modules/super-regex/index.js";
var maxIPv4Length = 15;
var maxIPv6Length = 45;
var options = {
  timeout: 400
};
function isIP(string) {
  return isMatch(ipRegex({ exact: true }), string.slice(0, maxIPv6Length), options);
}
function isIPv6(string) {
  return isMatch(ipRegex.v6({ exact: true }), string.slice(0, maxIPv6Length), options);
}
function isIPv4(string) {
  return isMatch(ipRegex.v4({ exact: true }), string.slice(0, maxIPv4Length), options);
}
function ipVersion(string) {
  return isIP(string) ? isIPv6(string) ? 6 : 4 : void 0;
}
export {
  ipVersion,
  isIP,
  isIPv4,
  isIPv6
};
//# sourceMappingURL=index.js.map
'
`;

snapshot[`compilers/esbuild > import map > supports scopes 1`] = `
'// lib/import_map/scopes.js
import "/lib/foo2.js";
//# sourceMappingURL=scopes.js.map
'
`;

snapshot[`compilers/esbuild > import map > supports aliasing 1`] = `
'// lib/import_map/aliases.js
import isip from "/node_modules/is-ip/index.js";
isip();
//# sourceMappingURL=aliases.js.map
'
`;

snapshot[`compilers/esbuild > Import relative module without extension 1`] = `
'// lib/import_relative_module_without_extension.js
import foo from "/lib/foo.js";
//# sourceMappingURL=import_relative_module_without_extension.js.map
'
`;

snapshot[`compilers/esbuild > Import absolute module without extension 1`] = `
'// lib/import_absolute_module_without_extension.js
import foo from "/lib/foo.js";
//# sourceMappingURL=import_absolute_module_without_extension.js.map
'
`;

snapshot[`compilers/esbuild > Import css module from JS 1`] = `
'// lib/styles.module.css
var e = document.querySelector("#_9095c7b8");
if (!e) {
  e = document.createElement("style");
  e.id = "_9095c7b8";
  document.head.appendChild(e);
  e.appendChild(document.createComment("/lib/styles.module.css"));
  e.appendChild(document.createTextNode(\`.myClass9095c7b8{color:pink}\`));
}
var styles_module_default = { "myClass": "myClass9095c7b8" };

// lib/import_css_module.js
console.log(styles_module_default);
//# sourceMappingURL=import_css_module.js.map
'
`;

snapshot[`compilers/esbuild > de-dupes side loaded ViewComponent default stylesheet - regular 1`] = `
'// app/components/basic_react_component.module.css
var basic_react_component_module_default = new Proxy({}, { get(target, prop, receiver) {
  if (prop in target || typeof prop === "symbol") {
    return Reflect.get(target, prop, receiver);
  } else {
    return prop + "454467aa";
  }
} });

// app/components/basic_react_component.jsx
import { jsx } from "/node_modules/react/jsx-runtime.js";
var basic_react_component_default = () => {
  return /* @__PURE__ */ jsx("div", { className: basic_react_component_module_default.base, children: "Basic React Component" });
};
export {
  basic_react_component_default as default
};
//# sourceMappingURL=basic_react_component.jsx.map
'
`;

snapshot[`compilers/esbuild > de-dupes side loaded ViewComponent default stylesheet - sidecar 1`] = `
'// app/components/basic_react/component.module.css
var component_module_default = new Proxy({}, { get(target, prop, receiver) {
  if (prop in target || typeof prop === "symbol") {
    return Reflect.get(target, prop, receiver);
  } else {
    return prop + "86cfc182";
  }
} });

// app/components/basic_react/component.jsx
import { jsx } from "/node_modules/react/jsx-runtime.js";
var component_default = () => {
  return /* @__PURE__ */ jsx("div", { className: component_module_default.base, children: "Basic React Component" });
};
export {
  component_default as default
};
//# sourceMappingURL=component.jsx.map
'
`;

snapshot[`compilers/esbuild > Import css from JS 1`] = `
'// app/views/layouts/application.css
var e = document.querySelector("#_16bcb75e");
if (!e) {
  e = document.createElement("style");
  e.id = "_16bcb75e";
  document.head.appendChild(e);
  e.appendChild(document.createComment("/app/views/layouts/application.css"));
  e.appendChild(document.createTextNode(\`body{color:pink}.name{color:#000}
\`));
}
//# sourceMappingURL=import_css.js.map
'
`;

snapshot[`compilers/esbuild > Import css from jsx 1`] = `
'// app/views/layouts/application.css
var e = document.querySelector("#_16bcb75e");
if (!e) {
  e = document.createElement("style");
  e.id = "_16bcb75e";
  document.head.appendChild(e);
  e.appendChild(document.createComment("/app/views/layouts/application.css"));
  e.appendChild(document.createTextNode(\`body{color:pink}.name{color:#000}
\`));
}
//# sourceMappingURL=import_css.jsx.map
'
`;

snapshot[`compilers/esbuild > ?bundle-all query string > js import 1`] = `
"// lib/bundle_all_import/three.js?bundle-all
console.log(3);

// lib/bundle_all_import/two.js?bundle-all
console.log(2);

// lib/bundle_all_import/one.js?bundle-all
console.log(1);
//# sourceMappingURL=index.js.map
"
`;

snapshot[`compilers/esbuild > ?bundle query string > js import 1`] = `
'// lib/foo.js?bundle
console.log("/lib/foo.js");

// lib/foo2.js?bundle
console.log("/lib/foo2.js");

// lib/foo3.js?bundle
console.log("/lib/foo3.js");

// node_modules/mypackage/index.js?bundle
console.log("node_modules/mypackage");

// lib/bundle_import/foo.js?bundle
console.log("bundle_import/foo.js");

// lib/component.jsx?bundle
import { jsx } from "/node_modules/react/jsx-runtime.js";
var Component = () => {
  return /* @__PURE__ */ jsx("div", { children: "Hello" });
};
var component_default = Component;

// lib/bundle_import/index.js
import emailRegex from "/url:https%3A%2F%2Fesm.sh%2Femail-regex";
import emailRegex2 from "/url:https%3A%2F%2Fesm.sh%2Femail-regex";
component_default();
emailRegex();
emailRegex2();
//# sourceMappingURL=index.js.map
'
`;

snapshot[`compilers/esbuild > ?bundle query string > css import 1`] = `
'@import "https://esm.sh/@radix-ui/colors@0.1.8/blue.css";
@import "https://esm.sh/@radix-ui/colors@0.1.8/gray.css";

/* lib/foo.css?bundle */
body {
  color: red;
}
body h1 {
  color: #00f;
}

/* node_modules/mypackage/styles.css?bundle */
.mypackage {
  color: red;
}

/* lib/styles.module.css?bundle */
.myClass9095c7b8 {
  color: pink;
}

/* lib/bundle_import.css */
header {
  color: pink;
}
/*# sourceMappingURL=bundle_import.css.map */
'
`;

snapshot[`compilers/esbuild > postcss > supports mixins 1`] = `
"/* lib/with_mixins.css */
a {
  color: red;
  font-size: 20px;
}
button {
  color: pink;
  font-size: 14px;
}
button span {
  margin: 10px;
}
/*# sourceMappingURL=with_mixins.css.map */
"
`;

snapshot[`compilers/esbuild > postcss > cssMixinPaths option 1`] = `
"/* lib/with_mixins_from_alternative_path.css */
a {
  color: red;
  font-size: 20px;
}
button {
  padding: 20px;
  font-size: 30px;
}
/*# sourceMappingURL=with_mixins_from_alternative_path.css.map */
"
`;
